internal/controller/clustersecurityintentbinding_controller.go

```diff
@@ -35,71 +39,184 @@ func (r *ClusterSecurityIntentBindingReconciler) Reconcile(ctx context.Context,
+				"ClusterNimbusPolicy.Name", req.Name, "ClusterSecurityIntentBinding.Name", req.Name,
+			)
+			return doNotRequeue()
 		}
 		logger.Error(err, "failed to get ClusterSecurityIntentBinding", "ClusterSecurityIntentBinding.Name", csib.Name)
-		return ctrl.Result{}, err
+		return requeueWithError(err)
 	}
+	logger.Info("ClusterSecurityIntentBinding found", "ClusterSecurityIntentBinding.Name", req.Name)
 
-	if csib.Status.Status == "" || csib.Status.Status == StatusPending {
-		csib.Status.Status = StatusCreated
-		if err := r.Status().Update(ctx, csib); err != nil {
-			logger.Error(err, "failed to update ClusterSecurityIntentBinding status", "ClusterSecurityIntentBinding.Name", csib.Name)
-			return ctrl.Result{}, err
-		}
-		// Let's re-fetch the ClusterSecurityIntentBinding CR after updating the status
-		// so that we have the latest state of the resource on the cluster.
-		if err := r.Get(ctx, req.NamespacedName, csib); err != nil {
-			logger.Error(err, "failed to re-fetch ClusterSecurityIntentBinding", "ClusterSecurityIntentBinding.Name", csib.Name)
-			return ctrl.Result{}, err
-		}
-		logger.Info("ClusterSecurityIntentBinding found", "ClusterSecurityIntentBinding.Name", csib.Name)
-		return ctrl.Result{}, nil
+	if err = r.createOrUpdateCwnp(ctx, logger, req); err != nil {
+		return requeueWithError(err)
 	}
 
-	bindingInfo := intentbinder.MatchAndBindIntentsGlobal(ctx, r.Client, csib)
+	if err = r.updateStatus(ctx, logger, req); err != nil {
+		return requeueWithError(err)
+	}
 
-	clusterNp, err := policybuilder.BuildClusterNimbusPolicy(ctx, r.Client, r.Scheme, bindingInfo)
-	if err != nil {
-		logger.Error(err, "failed to build ClusterNimbusPolicy")
-		return ctrl.Result{}, err
+	return doNotRequeue()
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ClusterSecurityIntentBindingReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.ClusterSecurityIntentBinding{}).
+		Owns(&v1.ClusterNimbusPolicy{}).
+		WithEventFilter(
+			predicate.Funcs{
+				CreateFunc: r.createFn,
+				UpdateFunc: r.updateFn,
+				DeleteFunc: r.deleteFn,
+			},
+		).
+		Complete(r)
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) createFn(createEvent event.CreateEvent) bool {
+	if _, ok := createEvent.Object.(*v1.ClusterNimbusPolicy); ok {
+		return false
 	}
+	return true
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) updateFn(updateEvent event.UpdateEvent) bool {
+	// TODO: Handle update event for ClusterNimbusPolicy update so that reconciler don't process it
+	// twice.
+	return updateEvent.ObjectOld.GetGeneration() != updateEvent.ObjectNew.GetGeneration()
+}
+
+func (r *ClusterSecurityIntentBindingReconciler) deleteFn(deleteEvent event.DeleteEvent) bool {
+	obj := deleteEvent.Object
+	if _, ok := obj.(*v1.ClusterSecurityIntentBinding); ok {
+		return true
+	}
+	return ownerExists(r.Client, obj)
```

Comment:
Since we get deletion event for CSIB and CWNP CRs, so what if an external CWNP or an owned CWNP gets deleted?

In that case we check whether the deleted CWNP is orphan if so ignore that delete event, otherwise recreate the deleted CWNP by checking the UID mentioned in its ownerref vs  the controller we found in the cluster.